server/helpers/requests.js

```diff
@@ -0,0 +1,379 @@
+const { UnreachableCaseError } = require('./errors')
+const {
+  getOriginFromUrl,
+  getWebAgentLink,
+  getWebDatasetLink
+} = require('./links')
+
+const DATASET_AUTHORIZATION_TYPES = {
+  CREATED: 'dataset.authorization.created',
+  DELETED: 'dataset.authorization.deleted',
+  REQUEST_CREATED: 'dataset.authorization_request.created',
+  REQUEST_CANCELLED: 'dataset.authorization_request.cancelled',
+  REQUEST_APPROVED: 'dataset.authorization_request.approved',
+  REQUEST_REJECTED: 'dataset.authorization_request.rejected',
+  INVITE_CREATED: 'dataset.authorization_invite.created',
+  INVITE_CANCELLED: 'dataset.authorization_invite.cancelled',
+  INVITE_APPROVED: 'dataset.authorization_invite.approved',
+  INVITE_REJECTED: 'dataset.authorization_invite.rejected'
+}
+
+const CONTRIBUTION_REQUEST_TYPES = {
+  CATALOG_CONTRIBUTE_CREATED: 'catalog.contribute_request.created',
+  CATALOG_CONTRIBUTE_CANCELLED: 'catalog.contribute_request.cancelled',
+  CATALOG_CONTRIBUTE_APPROVED: 'catalog.contribute_request.approved',
+  CATALOG_TRANSFER_CREATED: 'catalog.transfer_request.created',
+  CATALOG_TRANSFER_CANCELLED: 'catalog.transfer_request.cancelled',
+  CATALOG_TRANSFER_APPROVED: 'catalog.transfer_request.approved',
+  DATASET_CONTRIBUTE_CREATED: 'dataset.contribute_request.created',
+  DATASET_CONTRIBUTE_CANCELLED: 'dataset.contribute_request.cancelled',
+  DATASET_CONTRIBUTE_APPROVED: 'dataset.contribute_request.approved'
+}
+
+function formatLink(text, url) {
+  return `<${url}|${text}>`
+}
+
+function getExtraDetailsText({
+  header,
+  object,
+  headerLink
+}) {
+  const linkedHeader = headerLink
+    ? `*${formatLink(header, headerLink)}*\n`
+    : `*${header}*\n`
+  return Object.entries(object).reduce(
+    (text, [key, val]) => (val ? text + `*${key}*: _${val}_\n` : text),
+    linkedHeader
+  )
+}
+
+function getAuthLevelCopyText(authorizationLevel) {
+  switch (authorizationLevel) {
+    case 'DISCOVER':
+      return 'discover'
+    case 'READ':
+      return 'view'
+    case 'WRITE':
+      return 'edit'
+    case 'ADMIN':
+      return 'manage'
+    default:
+      throw new UnreachableCaseError(authorizationLevel)
+  }
+}
+
+function getDatasetAccessCopyText(
+  authorizationLevel,
+  dataset
+) {
+  return `_${getAuthLevelCopyText(authorizationLevel)}_ access to _${dataset}_`
+}
+
+function hasNonEmptyProperties(object) {
+  return Object.values(object).some(val => val)
+}
+
+function getAuthorizationSummaryText(eventBody) {
+  const origin = getOriginFromUrl(eventBody.resourceUrl)
```

Comment:
For the URLs that are built based on agentid/datasetid, we should probably use the same origin as the URLs that are sent in the webhook request. I'm currently just getting the origin from the other URLs that are sent in the webhook body, which feels slightly hacky.

We could add the origin as a property in the webhook body, but I'm not sure if it's necessary.

Reply:
Does `slack-app` not build any other full-qualified d.w urls?

Reply:
It does. In the other cases it gets the origin from the origin of the incoming request. Since we get URLs in the webhook bodies anyways, I feel like using the webhook body URLs to get the origin instead of the incoming request origin is a more consistent approach for this scenario.

I don't think configuring the Slack app with a single origin would work here, since the Slack app could receive requests from multiple origins once we get sites functionality, right?

Reply:
> the Slack app could receive requests from multiple origins once we get sites functionality

this is a problem we are going to need to solve for `notifications` as well, but if the Slack app is not currently doing this then let's not add that complexity

Your assumption that the origins will always match is probably fine but I agree that it feels hacky and there might not always be another url you can rely on.

I think looking at the origin of the incoming request feels a bit cleaner than relying on a property of the webhook itself, but I also worry that will stop working for Sites as well.

Ugh I'm going in circles. Let's just keep it this way for now.

Reply:
Sure, I'll leave a comment about it as well.